### PR TITLE
Allow content to be a promise and use resolved promise as the rows

### PR DESCRIPTION
**Pull request by BryanCrotaz**
Fixes https://github.com/Addepar/ember-table-addon/issues/16

### CODE REVIEW

File: addon/components/ember-table.js [hunk 1]
```diff
@@ -115,6 +115,40 @@ StyleBindingsMixin, ResizeHandlerMixin, {
 
   columnsFillTable: true,
 
+  _content: function(key, value) {
+    if (arguments.length > 1) {
+      // > 1 args = this is a `set`
+      return value;
+    } else {
+      // otherwise this is a `get`
+      var self = this;
```
Comment (azirbel) [inline addon/components/ember-table.js]: we typically use `_this` instead of `self`

File: addon/components/ember-table.js [hunk 2]
```diff
@@ -115,6 +115,40 @@ StyleBindingsMixin, ResizeHandlerMixin, {
...
+      value = null;
+
+      var content = this.get('content');
+      if (content.then)
+      {
+        // content is a promise
+        content.then(function(resolvedContent) {
+          // when the promise is resolved, set this property with the value
+          self.set("_content", resolvedContent);
```
Comment (azirbel) [inline addon/components/ember-table.js]: single quotes

File: addon/components/ember-table.js [hunk 3]
```diff
@@ -115,6 +115,40 @@ StyleBindingsMixin, ResizeHandlerMixin, {
 
   columnsFillTable: true,
 
+  _content: function(key, value) {
```
Comment (azirbel) [inline addon/components/ember-table.js]: I think the code would be more readable if you replaced the comments on lines 130, 132, 140 and 146 with a single comment above the `_content` computed property explaining the intention of this property. If this comment explains that `_content` exists to handle promises, then the code below is pretty clear without additional explanation

**Comment by azirbel** [comment]
This is great! Thanks for contributing a PR - people have wanted this behavior for a long time!

(1) please add a (very brief) description of this behavior to `documentation.hbs`
(2) please run `grunt dist` to generate the globals build

Other than that, LGTM. I'll merge once you address these comments.

**Comment by BryanCrotaz** [comment]
I'll make the changes you requested tomorrow.

I was surprised how easy it was - I've been studying the code for ages
trying to work out how to do it and when the penny dropped and I had
the code 5 mins later I was sure I'd missed something!

Couldn't find the unit tests to add this to...

**Comment by azirbel** [comment]
Oh, one more question: why do you return `null` instead of `[]` when the promise hasn't loaded?

We haven't set up tests yet ... it's next on the list after we release the Ember CLI version and make some minor API changes. For now, it's ok to skip writing them.

**Comment by BryanCrotaz** [comment]
Because the default value for content was null

On 13 May 2015 at 18:58, Alex Zirbel <notifications@github.com> wrote:

> Oh, one more question: why do you return null instead of [] when the
> promise hasn't loaded?
>
> We haven't set up tests yet ... it's next on the list after we release the
> Ember CLI version and make some minor API changes. For now, it's ok to skip
> writing them.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Addepar/ember-table-addon/pull/18#issuecomment-101759929>
> .
>



-- 
Bryan Crotaz
Managing Director
Silver Curve

**Comment by BryanCrotaz** [comment]
Can't find `documentation.hbs` in the repo?

**Comment by BryanCrotaz** [comment]
Tried to run `grunt dist` - what do I need to install first?

```
G:\Projects\ember-table-addon>grunt dist
'grunt' is not recognized as an internal or external command,
operable program or batch file.
```

**Comment by jrhe** [comment]
Sounds like you need to install grunt globally
`npm install -g grunt-cli`

**Comment by BryanCrotaz** [comment]
Didn't work. Tried installing globally then when it complained, tried locally:

```
G:\Projects\ember-table-addon>npm install -g grunt-cli
g:\nodejs\grunt -> g:\nodejs\node_modules\grunt-cli\bin\grunt
grunt-cli@0.1.13 g:\nodejs\node_modules\grunt-cli
├── resolve@0.3.1
├── nopt@1.0.10 (abbrev@1.0.5)
└── findup-sync@0.1.3 (lodash@2.4.2, glob@3.2.11)

G:\Projects\ember-table-addon>grunt dist
grunt-cli: The grunt command line interface. (v0.1.13)

Fatal error: Unable to find local grunt.

If you're seeing this message, either a Gruntfile wasn't found or grunt
hasn't been installed locally to your project. For more information about
installing and configuring grunt, please see the Getting Started guide:

http://gruntjs.com/getting-started

G:\Projects\ember-table-addon>npm install grunt-cli
npm WARN prefer global grunt-cli@0.1.13 should be installed with -g
grunt-cli@0.1.13 node_modules\grunt-cli
├── resolve@0.3.1
├── nopt@1.0.10 (abbrev@1.0.5)
└── findup-sync@0.1.3 (lodash@2.4.2, glob@3.2.11)

G:\Projects\ember-table-addon>grunt dist
grunt-cli: The grunt command line interface. (v0.1.13)

Fatal error: Unable to find local grunt.

If you're seeing this message, either a Gruntfile wasn't found or grunt
hasn't been installed locally to your project. For more information about
installing and configuring grunt, please see the Getting Started guide:

http://gruntjs.com/getting-started

```

**Comment by jrhe** [comment]
Try doing `npm install` again. Grunt is a dependency so it should be installed if you did a npm install. If not try removing the node modules folder and trying again.

**Comment by azirbel** [comment]
https://github.com/Addepar/ember-table-addon/blob/master/tests/dummy/app/templates/documentation.hbs; if you still have trouble with `grunt dist` after the last suggestion from @jrhe then don't worry about it; I can run it for you

**Comment by BryanCrotaz** [comment]
Ah - I've forked  addon, which doesn't appear to have blob in it

**Comment by BryanCrotaz** [comment]
Ok, I've run `grunt dist` - what's the output supposed to be? Sourcetree isn't seeing any file changes

**Comment by BryanCrotaz** [comment]
Found the documentation - wasn't expecting to find docs under the tests folder!

**Comment by azirbel** [comment]
Hehe, it's a debatable decision but we've put the demo app in `tests/dummy`, which is a standard place to put a little showcase/hello-world app.

`grunt dist` should create `dist` files: https://github.com/Addepar/ember-table-addon/tree/master/dist.

**Comment by BryanCrotaz** [comment]
Ran again:
```
G:\Projects\ember-table-addon>grunt dist
Running "replace:global_version" (replace) task

Running "replace:overview_page" (replace) task

Running "broccoli:dist:build" (broccoli) task
```
No output in dist

**Comment by azirbel** [comment]
Weird! I do get dist files when I run; see https://github.com/Addepar/ember-table-addon/tree/azirbel/pr/18.

Don't worry about it. I'll run it myself and amend your commit (you will still be listed as the author).

Everything looks good to me - are you done making changes?

**Comment by BryanCrotaz** [comment]
Yes

**Comment by azirbel** [comment]
Squashed commits, added `dist` files, and merged via https://github.com/Addepar/ember-table-addon/commit/a5190e815e08466748c3422c1d15ca76ba4aa0ba. Thanks @BryanCrotaz!

**Pull request closed without merging**